File: Factorial.cpp
#include <iostream>
#include <string>

template<int N>
struct Factorial {
    static const int value = N * Factorial<N-1>::value;
};

template<>
struct Factorial<1> {
    static const int value = 1;
};

int main() {
    std::cout << Factorial<5>::value << "\n";
}

File: Factorial.java
class Factorial {
    public static void main(String[] args) {
        int i = Integer.parseInt(args[0]);
        System.out.println(factorial(i));
    }

    // public static int factorial(int i) {
    //     int result = i;
    //     if (i == 0) return result;
    //     else {
    //         while (i != 1) {
    //             result *= --i;
    //         }
    //         return result;
    //     }
    // }

    public static int factorial(int i) {
        if (i == 1) return i;
        else return i * factorial(i-1);
    }
}